src/main/tutorials/calculate-metrics/jqassistant/ood-metrics.adoc

```diff
@@ -0,0 +1,315 @@
+----
+MATCH (p:Package {fqn:"your.project.oodmetrics.b", ca:3, ce:3, instability:0.5})
+RETURN p
+----
+
+[[ood-metrics:InstabilityPackageC]]
+.Instability Package c
+[source,cypher,role=concept,requiresConcepts="ood-metrics:Instability",severity=major]
+----
+MATCH (p:Package {fqn:"your.project.oodmetrics.c", ca:1, ce:3, instability:0.75})
+RETURN p
+----
+
+[[ood-metrics:InstabilityPackageD]]
+.Instability Package d
+[source,cypher,role=concept,requiresConcepts="ood-metrics:Instability",severity=major]
+----
+MATCH (p:Package {fqn:"your.project.oodmetrics.d", ca:0, ce:2, instability:1})
+RETURN p
+----
+
+[[ood-metrics:InstabilityPackageE]]
+.Instability Package e
+[source,cypher,role=concept,requiresConcepts="ood-metrics:Instability",severity=major]
+----
+MATCH (p:Package {fqn:"your.project.oodmetrics.e", ca:2, ce:0, instability:0})
+RETURN p
+----
+
+[[ood-metrics:InstabilityTests]]
+[role=group,includesConcepts="ood-metrics:InstabilityPackageA,ood-metrics:InstabilityPackageB,ood-metrics:InstabilityPackageC,ood-metrics:InstabilityPackageD,ood-metrics:InstabilityPackageE"]
+
+=== Abstractness Metrics
+
+The abstractness is the ratio of abstract classes and interfaces to the total number of classes and interfaces in a package.
+
+==== Na (number abstracts)
+
+The number of abstract classes and interfaces in this component.
+
+[[ood-metrics:NumberAbstracts]]
+.Computes the number of abstract classes and interfaces for all packages
+[source,cypher,role=concept]
+----
+MATCH (p:Package) WITH p
+OPTIONAL MATCH (p)-[:CONTAINS]->(ac:Java:Type:Class {abstract:true}) WITH p, COUNT(ac) AS numberAbstractClasses
+OPTIONAL MATCH (p)-[:CONTAINS]->(i:Java:Type:Interface) WITH p, numberAbstractClasses, COUNT(i) AS numberInterfaces
+SET p.na = COALESCE(numberAbstractClasses + numberInterfaces, 0)
+RETURN p
+----
+
+==== Nc (number classes)
+
+The number of all classes and interfaces in this component.
+
+[[ood-metrics:NumberClasses]]
+.Computes the number of all classes and interfaces for all packages
+[source,cypher,role=concept]
+----
+MATCH (p:Package)
+OPTIONAL MATCH (p)-[:CONTAINS]->(c:Java:Type)
+WITH p, COUNT(c) AS numberClasses
+SET p.nc = COALESCE(numberClasses, 0)
+RETURN p
+----
+
+==== A (abstractness)
+
+A = Na / Nc +
+The ratio of abstract classes and interfaces to the number of all classes and interfaces per package.
+This metric has the range [0, 1].
+A=0 indicates that this package has no abstract classes or interfaces.
+A=1 indicates that this package only contains abstract classes or interfaces.
+
+[[ood-metrics:Abstractness]]
+.Computes the abstractness for all packages
+[source,cypher,role=concept,requiresConcepts="ood-metrics:NumberAbstracts,ood-metrics:NumberClasses"]
+----
+MATCH (p:Package)
+WHERE p.nc > 0
+SET p.abstractness = toFloat(p.na) / p.nc
+RETURN p
+----
+
+==== Example
+
+.OO-Metrics Abstractness
+image::OO-Metrics_Abstractness.png[OO-Metrics Abstractness]
+
+==== Test cases
```

Comment:
The test cases are confusing for me? What should be tested here? Wouldn't it be better for the reader see how he can define constraints to ensure that specific thresholds won't be violated?